cocotb/bfmgen.py

```diff
@@ -0,0 +1,364 @@
+            p = imp.signature[pi]
+            if p.ptype.s:
+                bfm_import_calls += "                  longint p" + str(pi) + " = cocotb_bfm_get_si_param(bfm_id);\n"
+            else:
+                bfm_import_calls += "                  longint unsigned p" + str(pi) + " = cocotb_bfm_get_ui_param(bfm_id);\n"
+                
+        bfm_import_calls += "                  " + imp.T.__name__  + "(\n"
+        for pi in range(len(imp.signature)):
+            bfm_import_calls += "                      p" + str(pi)
+
+            if pi+1 < len(imp.signature):
+                bfm_import_calls += ","
+            bfm_import_calls += "\n"
+        bfm_import_calls += "                      );\n"
+        bfm_import_calls += "              end\n"
+        
+    bfm_export_tasks = ""
+    for i in range(len(info.export_info)):
+        exp = info.export_info[i]
+        bfm_export_tasks += "    task " + exp.T.__name__ + "("
+        for j in range(len(exp.signature)):
+            p = exp.signature[j]
+            bfm_export_tasks += p.ptype.vl_type() + " " + p.pname
+            if j+1 < len(exp.signature):
+                bfm_export_tasks += ", "
+        bfm_export_tasks += ");\n"
+        bfm_export_tasks += "    begin\n"
+        bfm_export_tasks += "        cocotb_bfm_begin_msg(bfm_id, " + str(i) + ");\n"
+        for p in exp.signature:
+            if p.ptype.s:
+                bfm_export_tasks += "        cocotb_bfm_add_si_param(bfm_id, " + p.pname + ");\n"
+            else:
+                bfm_export_tasks += "        cocotb_bfm_add_ui_param(bfm_id, " + p.pname + ");\n"
+            
+        bfm_export_tasks += "        cocotb_bfm_end_msg(bfm_id);\n"
+        bfm_export_tasks += "    end\n"
+        bfm_export_tasks += "    endtask\n"
+        
+    
+    impl_param_m = {
+        "bfm_name" : bfm_name,
+        "bfm_classname" : info.T.__module__ + "." + info.T.__qualname__,
+        "bfm_import_calls" : bfm_import_calls,
+        "bfm_export_tasks" : bfm_export_tasks
+        }
+    
+    cocotb_bfm_api_impl = '''
+    int          bfm_id;
+    
+    import "DPI-C" context function int cocotb_bfm_claim_msg(int bfm_id);
+    import "DPI-C" context function longint cocotb_bfm_get_si_param(int bfm_id);
+    import "DPI-C" context function longint unsigned cocotb_bfm_get_ui_param(int bfm_id);
+    import "DPI-C" context function void cocotb_bfm_begin_msg(int bfm_id, int msg_id);
+    import "DPI-C" context function void cocotb_bfm_add_si_param(int bfm_id, longint v);
+    import "DPI-C" context function void cocotb_bfm_add_ui_param(int bfm_id, longint unsigned v);
+    import "DPI-C" context function void cocotb_bfm_end_msg(int bfm_id);
+    
+    task automatic ${bfm_name}_process_msg();
+        int msg_id = cocotb_bfm_claim_msg(bfm_id);
+        case (msg_id)
+${bfm_import_calls}
+        default: begin
+            $display("Error: BFM %m received unsupported message with id %0d", msg_id);
```

Comment:
I like to print the simulation time in such cases, but since you ``$finish`` anyway, it's going to be implicitly done for most if not all simulators.